### PR TITLE
Land v1 HTTP / cross-JVM transport

### LINKED ISSUE #9: [v1] Implement HTTP / cross-JVM mode
Context: tracked from the v1 / Phase 1 checklist in the Compose Desktop automator gist.

Reference:
- Gist: https://gist.github.com/rock3r/8e520bb3fe8fe5886367d5e22cefbab8

Goal: deliver the v1 transport layer so the automator works cross-JVM as well as in-process.

Checklist:
- [ ] Implement transport DTOs / serialization for the public automator model
- [ ] Implement the embedded HTTP server
- [ ] Implement the HTTP client (`ComposeAutomator.http(...)`)
- [ ] Reach API parity between in-process and HTTP modes
- [ ] Add contract tests for HTTP payloads / request-response boundaries

### PR DESCRIPTION
**Pull request by rock3r**
## Summary
- Adds the opt-in HTTP transport: a `ComposeAutomator` running in one JVM can now be driven from another via `ComposeAutomator.http(host, port)`.
- Server side: `Application.installSpectreRoutes(automator, basePath = \"/spectre\")` Ktor extension. Mounts `GET /windows`, `GET /nodes` (with optional `?testTag=`), `POST /click`, `POST /typeText`, `GET /screenshot`.
- Client side: `HttpComposeAutomator` (AutoCloseable) with the matching v1 surface. Created via the companion extension `ComposeAutomator.Companion.http(...)`.
- DTOs in `dev.sebastiano.spectre.server.dto` are kotlinx-serialization data classes pinning the wire contract.
- Pre-existing API leak fixed: `core` now declares `compose-ui` as `api` (was `implementation`) since `AutomatorNode.boundsInWindow` returns `androidx.compose.ui.geometry.Rect`.

## v1 scope
The endpoints land the most-used queries and actions. The following stay in-process-only and are documented in `server/README.md`:
- `registerIdlingResource` / `waitForIdle` / `waitForVisualIdle` (JVM objects + long-poll semantics out of scope for v1 transport)
- `withTracing` (needs a JVM `Tracer` instance the server can't accept across processes)
- `printTree` (debug helper; not transport-essential)

Closes #9.

## Test plan
- [x] `./gradlew check` — all modules green
- [x] `DtoSerializationTest` (5) — every DTO round-trips through kotlinx-serialization, including default-only and fully-populated variants
- [x] `SpectreServerRoundTripTest` (7) — Ktor `testApplication` exercises the real route handlers against a headless `ComposeAutomator`: list shapes, status codes, body decoding, custom base path, 404 for unknown node keys
- [ ] Live cross-JVM smoke against `sample-desktop` is part of #12 (sample harness coverage)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: server/src/main/kotlin/dev/sebastiano/spectre/server/SpectreServer.kt [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+                windows = automator.windows.mapIndexed { index, window -> window.toDto(index) }
+            )
+        call.respond(response)
+    }
+
+    get("/nodes") {
+        val testTag = call.request.queryParameters["testTag"]
+        val nodes =
+            if (testTag != null) {
+                automator.findByTestTag(testTag)
```
[the remark this answers is not in the document]
Comment (rock3r) [reply server/src/main/kotlin/dev/sebastiano/spectre/server/SpectreServer.kt:76]: Fixed in 3a83f88 — `/nodes` and `/click` handlers now call `refreshWindows()` first, matching the documented "remote callers see fresh state per request" semantics.

File: server/src/main/kotlin/dev/sebastiano/spectre/server/SpectreServer.kt [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+ * [automator]. All routes live under [basePath] (default `/spectre`).
+ *
+ * The route surface is intentionally a v1 subset of the public automator API — windows, nodes by
+ * tag, click, typeText, screenshot — enough to drive cross-JVM smoke tests. Advanced features
+ * (idling resources, withTracing, waitForVisualIdle) are in-process only because they either
+ * require live JVM objects (Tracer, IdlingResource) or stateful long-poll semantics that are out of
+ * scope for the v1 transport.
+ */
+fun Application.installSpectreRoutes(automator: ComposeAutomator, basePath: String = "/spectre") {
+    install(ContentNegotiation) { json() }
```
[the remark this answers is not in the document]
Comment (rock3r) [reply server/src/main/kotlin/dev/sebastiano/spectre/server/SpectreServer.kt]: Fixed in 3a83f88 — `installSpectreRoutes` now checks `pluginOrNull(ContentNegotiation)` and only installs when absent. Added a regression test that pre-installs the plugin on the host application.

File: server/src/main/kotlin/dev/sebastiano/spectre/server/dto/Conversions.kt [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        isFocused = isFocused,
+        isDisabled = isDisabled,
+        isSelected = isSelected,
+        boundsInWindow =
+            RectangleDto(
+                x = boundsInWindow.left.toDouble(),
+                y = boundsInWindow.top.toDouble(),
+                width = (boundsInWindow.right - boundsInWindow.left).toDouble(),
+                height = (boundsInWindow.bottom - boundsInWindow.top).toDouble(),
+            ),
```
[the remark this answers is not in the document]
Comment (rock3r) [reply server/src/main/kotlin/dev/sebastiano/spectre/server/dto/Conversions.kt:39]: Fixed in 3a83f88 — `AutomatorNode.toDto()` snapshots `boundsInWindow` into a single local val before reading its components, so layout shifts mid-conversion can no longer produce inconsistent rectangles.

File: server/src/main/kotlin/dev/sebastiano/spectre/server/dto/Conversions.kt [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+internal fun Rectangle.toDto(): RectangleDto =
+    RectangleDto(
+        x = x.toDouble(),
+        y = y.toDouble(),
+        width = width.toDouble(),
+        height = height.toDouble(),
+    )
+
+internal fun RectangleDto.toAwt(): Rectangle =
+    Rectangle(x.toInt(), y.toInt(), width.toInt(), height.toInt())
```
[the remark this answers is not in the document]
Comment (rock3r) [reply server/src/main/kotlin/dev/sebastiano/spectre/server/dto/Conversions.kt]: Fixed in 3a83f88 — removed `RectangleDto.toAwt()`, no callers existed.

File: server/src/main/kotlin/dev/sebastiano/spectre/server/SpectreServer.kt [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+    get("/nodes") {
+        val testTag = call.request.queryParameters["testTag"]
+        val nodes =
+            if (testTag != null) {
+                automator.findByTestTag(testTag)
+            } else {
+                automator.allNodes()
+            }
+        call.respond(NodesResponse(nodes = nodes.map { it.toDto() }))
```
[the remark this answers is not in the document]
Comment (rock3r) [reply server/src/main/kotlin/dev/sebastiano/spectre/server/SpectreServer.kt:80]: Fixed in 3a83f88 — same change as the Codex thread above. `/nodes` and `/click` handlers now refresh windows first.

File: server/src/main/kotlin/dev/sebastiano/spectre/server/HttpComposeAutomator.kt [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+ * queries and actions; advanced features (idling resources, `withTracing`, `waitForVisualIdle`)
+ * remain in-process only and are documented as such.
+ *
+ * The instance owns its [HttpClient] and must be `close()`d to release pooled connections. Use `use
+ * { ... }` from `kotlin.AutoCloseable` for scoped lifecycles.
+ */
+class HttpComposeAutomator
+internal constructor(private val baseUrl: String, engine: HttpClientEngine) : AutoCloseable {
+
+    private val client: HttpClient = HttpClient(engine) { install(ContentNegotiation) { json() } }
```
[the remark this answers is not in the document]
Comment (rock3r) [reply server/src/main/kotlin/dev/sebastiano/spectre/server/HttpComposeAutomator.kt]: Fixed in 7a86141 — switched to `HttpClient(CIO) { ... }` (engine factory form), which makes the client own the engine. `close()` now shuts down both the client and the CIO connection pool / selector threads. Dropped the unused internal `engine` parameter from `create()`.

File: server/src/main/kotlin/dev/sebastiano/spectre/server/SpectreServer.kt [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    post("/click") {
+        automator.refreshWindows()
+        val request = call.receive<ClickRequest>()
+        val key = NodeKey.parse(request.nodeKey)
```
[the remark this answers is not in the document]
Comment (rock3r) [reply server/src/main/kotlin/dev/sebastiano/spectre/server/SpectreServer.kt]: Fixed in 7a86141 — `NodeKey.parse` is now wrapped in try/catch and surfaces malformed input as `400 Bad Request` with the parse error message. Added a regression test (`click against a malformed node key returns 400`).

File: server/src/main/kotlin/dev/sebastiano/spectre/server/HttpComposeAutomator.kt [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+        val response = client.get("$baseUrl/screenshot").body<ScreenshotResponse>()
+        val bytes = Base64.getDecoder().decode(response.pngBase64)
+        return checkNotNull(ImageIO.read(ByteArrayInputStream(bytes))) {
+            "Server returned an image we could not decode"
+        }
+    }
+
+    override fun close() {
+        client.close()
+    }
```
[the remark this answers is not in the document]
Comment (rock3r) [reply server/src/main/kotlin/dev/sebastiano/spectre/server/HttpComposeAutomator.kt:89]: Fixed in 7a86141 — same change as the Codex thread above. The `HttpClient(CIO)` factory form ties the engine lifecycle to the client.

File: server/src/main/kotlin/dev/sebastiano/spectre/server/HttpComposeAutomator.kt [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+
+    companion object {
+
+        /** Default port suggested by the gist's HTTP example. */
+        const val DEFAULT_PORT: Int = 9274
+
+        internal fun create(host: String, port: Int, basePath: String): HttpComposeAutomator =
+            HttpComposeAutomator(baseUrl = "http://$host:$port$basePath")
```
[the remark this answers is not in the document]
Comment (rock3r) [reply server/src/main/kotlin/dev/sebastiano/spectre/server/HttpComposeAutomator.kt]: Fixed in b3d314a — extracted `HttpComposeAutomator.normaliseBaseUrl(host, port, basePath)` that trims slashes and prefixes exactly one. Added unit tests covering all the variants (`/spectre`, `spectre`, `/spectre/`, multi-segment paths, empty path).

File: server/src/main/kotlin/dev/sebastiano/spectre/server/dto/Conversions.kt [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+        isDisabled = isDisabled,
+        isSelected = isSelected,
+        boundsInWindow =
+            RectangleDto(
+                x = rect.left.toDouble(),
+                y = rect.top.toDouble(),
+                width = (rect.right - rect.left).toDouble(),
+                height = (rect.bottom - rect.top).toDouble(),
+            ),
+        boundsOnScreen = boundsOnScreen.toDto(),
```
[the remark this answers is not in the document]
Comment (rock3r) [reply server/src/main/kotlin/dev/sebastiano/spectre/server/dto/Conversions.kt]: Fixed in b3d314a — added `AutomatorNode.bothBounds()` (returns `BothBounds`) which reads both forms in a single `readOnEdt` block. The DTO conversion now uses it, so the two rectangles are guaranteed to reflect the same underlying layout state.

File: server/build.gradle.kts [hunk 11]
```diff
@@ -1,12 +1,35 @@
...
 dependencies {
-    implementation(projects.core)
+    api(projects.core)
+    api(libs.kotlinx.serialization.json)
+
+    // Server runtime is `implementation` so consumers hosting the automator inherit it via
+    // `api(projects.core)` + `implementation(projects.server)` only when they actually want the
+    // server side. The `installSpectreRoutes` extension and `HttpComposeAutomator` client are
+    // both reachable transitively from the public API.
+    implementation(libs.ktor.server.core)
```
[the remark this answers is not in the document]
Comment (rock3r) [reply server/build.gradle.kts]: Fixed in 6cf7191 — promoted `ktor-server-core` from `implementation` to `api` so consumers can compile against `installSpectreRoutes` without re-declaring Ktor.

File: server/src/main/kotlin/dev/sebastiano/spectre/server/SpectreServer.kt [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+ * (idling resources, withTracing, waitForVisualIdle) are in-process only because they either
+ * require live JVM objects (Tracer, IdlingResource) or stateful long-poll semantics that are out of
+ * scope for the v1 transport.
+ */
+fun Application.installSpectreRoutes(automator: ComposeAutomator, basePath: String = "/spectre") {
+    // Hosts may already have ContentNegotiation installed for their own routes; Ktor throws a
+    // duplicate-plugin exception on a second install. Only install when absent so the route
+    // mount is safe to call from any application configuration.
+    if (pluginOrNull(ContentNegotiation) == null) {
+        install(ContentNegotiation) { json() }
```
[the remark this answers is not in the document]
Comment (rock3r) [reply server/src/main/kotlin/dev/sebastiano/spectre/server/SpectreServer.kt:54]: Documented in 6cf7191 — Ktor`s plugin model does not allow merging converters into an already-installed `ContentNegotiation`, and re-installing throws a duplicate-plugin exception. The KDoc on `installSpectreRoutes` now spells out the contract: when the host pre-installs CN, the host must include a JSON converter; otherwise routes fail at request time. When CN is absent, we install with kotlinx JSON automatically.

File: server/build.gradle.kts [hunk 13]
```diff
@@ -1,12 +1,35 @@
...
+    implementation(libs.ktor.server.cio)
```
[the remark this answers is not in the document]
Comment (rock3r) [reply server/build.gradle.kts]: Fixed in 6cf7191 — removed the unused `ktor-server-cio` dependency. `installSpectreRoutes` is engine-agnostic, so consumers bring their own server engine.
